### PR TITLE
Add refs::{isposinf, isneginf}

### PR DESCRIPTION
**Pull request by zasdfgbnm**

### CODE REVIEW

File: torch/_prims/__init__.py
Comment (mruberry) [inline torch/_prims/__init__.py:699]: Nice fix

File: torch/_refs/__init__.py
Comment (mruberry) [inline torch/_refs/__init__.py:469]: logical_or?

File: torch/_refs/__init__.py
Comment (mruberry) [inline torch/_refs/__init__.py:470]: again -- logical and not bitwise here?

File: torch/_refs/__init__.py
Comment (mruberry) [inline torch/_refs/__init__.py:481]: `gt(a, 0)` -- if `gt` doesn't accept a scalar then please add a TODO comment here and to the gt ref

File: torch/_refs/__init__.py
Comment (mruberry) [inline torch/_refs/__init__.py:492]: Same here with `lt(a, 0)`

File: torch/_refs/__init__.py
Comment (mruberry) [inline torch/_refs/__init__.py:604]: Existential prim modeling question:

One of the things we're trying to do with prims is make them orthogonal to each other. Of course we don't want to take this to extremes, but would it make more sense for `prims.real` and `prims.imag` to only work on complex tensors? Then `refs.real` would call `prims.view_of` for non-complex tensors, and only call `prims.real` on complex tensors.

**Review by mruberry — COMMENTED** [review]
Looks great @zasdfgbnm -- I made a few inline suggestions. Curious to hear your thoughts on prim modeling and swapping a scalar for instantiating a full tensor in a couple cases

File: torch/_refs/__init__.py
Comment (zasdfgbnm) [reply torch/_refs/__init__.py]: fixed

File: torch/_refs/__init__.py
Comment (zasdfgbnm) [reply torch/_refs/__init__.py]: There is no `logical_not`, so I keep the `bitwise_not` and added a TODO.

File: torch/_refs/__init__.py
Comment (zasdfgbnm) [reply torch/_refs/__init__.py]: I think this makes sense. Fixed.

File: torch/_prims/__init__.py
Comment (zasdfgbnm) [inline torch/_prims/__init__.py:359]: This will be useful for `real`, `imag`, `conj`, `angle`, `sgn`

File: torch/_refs/__init__.py
Comment (zasdfgbnm) [reply torch/_refs/__init__.py]: fixed

**Comment by zasdfgbnm** [comment]
All review comments resolved

File: torch/_refs/__init__.py
Comment (mruberry) [inline torch/_refs/__init__.py:493]: Should this just be 

```
return eq(a, float('inf'))
```

?

File: torch/_refs/__init__.py
Comment (mruberry) [inline torch/_refs/__init__.py:504]:
```
return eq(a, float('-inf'))
```

?

File: torch/_refs/__init__.py
Comment (mruberry) [inline torch/_refs/__init__.py:618]: Nice detail that this returns `a` itself and not a view of `a`

**Review by mruberry — APPROVED** [review]
Awesome! Thanks @zasdfgbnm!

I made two suggestions inline below that I'd appreciate your thoughts on before merging. Basically if `isposinf` and `isneginf` should just use `eq`.

File: torch/_refs/__init__.py
Comment (zasdfgbnm) [reply torch/_refs/__init__.py:493]: I don't think that will work, see
```python
t = torch.zeros(5, dtype=torch.int)
torch.eq(t, float('inf'))
```
```
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
Input In [5], in <cell line: 2>()
      1 t = torch.zeros(5, dtype=torch.int)
----> 2 torch.eq(t, float('inf'))

RuntimeError: value cannot be converted to type int without overflow
```

File: torch/_refs/__init__.py
Comment (zasdfgbnm) [reply torch/_refs/__init__.py:504]: See https://github.com/pytorch/pytorch/pull/78655/files#r889423851

File: torch/_refs/__init__.py
Comment (ngimel) [reply torch/_refs/__init__.py:493]: That seems like a bug in `eq` (and friends)

Comment (mruberry) [reply torch/_refs/__init__.py:493]: Also for integer tensors we should short-circuit to return full_like False and not try this comparison

File: torch/_refs/__init__.py
Comment (mruberry) [reply torch/_refs/__init__.py:504]: You're right, sorry, but I think we can just add a short-circuit for integer (and boolean) tensors

File: torch/_refs/__init__.py
Comment (mruberry) [reply torch/_refs/__init__.py:493]: Also for complex tensors isn't the definition of complex inf that either part is complex? How does that work for isposinf? Should we error on complex values to this function, or if any part is positive inf is that sufficient?

Comment (ngimel) [reply torch/_refs/__init__.py:493]: `isposinf` is not supported for complex inputs.

Comment (mruberry) [reply torch/_refs/__init__.py:493]: We should probably add an explicit check, then

**Comment by zasdfgbnm** [comment]
Just returned from vacation, will resolve review comments and merge.

File: torch/_refs/__init__.py
Comment (zasdfgbnm) [reply torch/_refs/__init__.py]: check for complex and short-circuit for integers added.

File: torch/_refs/__init__.py
Comment (zasdfgbnm) [reply torch/_refs/__init__.py]: check for complex and short-circuit for integers added.

**Comment by mruberry** [comment]
> Just returned from vacation, will resolve review comments and merge.

Welcome back, @zasdfgbnm! Hope to see you later this week!

File: torch/_refs/__init__.py
Comment (zasdfgbnm) [reply torch/_refs/__init__.py]: Looks like `ref_executor` does not support that. Looking deep into it.

File: torch/_prims/executor.py
Comment (zasdfgbnm) [inline torch/_prims/executor.py:10]: @mruberry Please review my latest change in this file

Comment (mruberry) [reply torch/_prims/executor.py:10]: Looks OK to me but I'd sanity check with @IvanYashchuk and/or @kevinstephano

File: torch/_prims/executor.py
Comment (kevinstephano) [inline torch/_prims/executor.py:71]: I was entertaining the idea of removing the need to call `add_input` all together since if you are defining a tensor or scalar, it is an input.  Therefore, NVFuser would just handle that accounting.

Comment (IvanYashchuk) [reply torch/_prims/executor.py:71]: Isn't `FusionDefinition.add_input` required to register the input as an input to the fusion? So that `fusion.execute` would know how to map inputs of the `execute` function to the inputs registered with `add_input`.
There are cases when defined tensor or scalar are not inputs to `fusion.execute` like
```py
def add2(fd, a):
    two_constant = fd.define_constant(2.0)
    fd.Ops.add(a, two_constant)
```

Comment (kevinstephano) [reply torch/_prims/executor.py:71]: There are 3 things that get defined separate from ops and those are Tensors, Scalars, and Constants.  Tensors and Scalars can result in fusion inputs but Constants do not.

In the C++ backend, the call for adding an input would not go away but, as a convenience, `define_tensor` and `define_scalar` could be augmented to additionally call the fusion objects `add_input` instead of having separate calls to `define_tensor` and then `add_input` at the `FusionDefinition` level.

I can't think of a situation where we define Tensors or Scalars where they are not inputs since every where else operations define the intermediates.  It's possible I am missing a case, though.

File: torch/_prims/executor.py
Comment (IvanYashchuk) [inline torch/_prims/executor.py:90]: Could you please clarify what function requires these changes?

**Review by IvanYashchuk — COMMENTED** [review]
Hey @zasdfgbnm, thank you for the proposed changes to the `executor` function! I've got a few questions for better understanding.

File: torch/_prims/executor.py
Comment (IvanYashchuk) [reply torch/_prims/executor.py:71]: Is there a concept of constant tensor in nvFuser?

Here's a case that I think is easy to find in users' scripts that fails with our current setup (tested on this branch):
```
In [1]: import torch

In [2]: from torch.fx.experimental.proxy_tensor import make_fx

In [3]: from torch._prims.executor import execute

In [4]: def func(a):
   ...:     b = torch.ones(a.shape, device='cuda')
   ...:     return torch._prims.add(a, b)
   ...:

In [5]: a = torch.randn(3, 3, device='cuda')

In [6]: gm = make_fx(func)(a)

In [7]: gm.graph.print_tabular()
opcode         name               target             args                      kwargs
-------------  -----------------  -----------------  ------------------------  --------
placeholder    a_1                a_1                ()                        {}
get_attr       _tensor_constant0  _tensor_constant0  ()                        {}
call_function  add                prims.add          (a_1, _tensor_constant0)  {}
output         output             output             (add,)                    {}

In [8]: execute(gm, a, executor="aten")
Out[8]:
tensor([[-0.0409,  1.5461,  0.0905],
        [ 1.4936,  0.8956,  1.2108],
        [ 1.1867,  1.8426,  1.7041]], device='cuda:0')

In [9]: execute(gm, a, executor="nvfuser")
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
File ~\dev\pytorch\torch\_prims\executor.py:79, in execute(gm, executor, *args, **kwargs)
     77         fd.add_output(o)
---> 79 nv_results = fusion.execute(
     80     tuple(arg for arg in args if isinstance(arg, torch.Tensor))
     81 )
     83 results = []

RuntimeError: thread_predicates_.find(tv_inp) != thread_predicates_.end() INTERNAL ASSERT FAILED at "C:\\Users\\Ivan\\dev\\pytorch\\torch\\csrc\\jit\\codegen\\cuda\\lower_thread_predicate.cpp":224, please report a bug to PyTorch. Thread predicate map was not initialized, couldn't find T1_l[ iblockIdx.x36{( ceilDiv(( ceilDiv(( ceilDiv(( i3 * i4 ), 128) ), 1) ), 1) )}, iUS37{1}, iS35{1}, ithreadIdx.x33{128} ]
```

File: torch/_prims/executor.py
Comment (IvanYashchuk) [reply torch/_prims/executor.py:45]: This change is required and all other points are just for discussion and shouldn't hold this PR from being merged if tests pass.

File: torch/_prims/executor.py
Comment (kevinstephano) [reply torch/_prims/executor.py:71]: There isn't a constant Tensor because NVFuser doesn't allocate memory.  If it did support an operation like `torch.ones()`, it would likely be through an operator similar to `randLike()` instead of via a constant definition.

File: torch/_prims/executor.py
Comment (zasdfgbnm) [reply torch/_prims/executor.py:49]: I think it is just python scalars.

File: torch/_prims/executor.py
Comment (zasdfgbnm) [reply torch/_prims/executor.py:90]: This is used to handle constant tensors in the output, such as the non-floating and non-complex case of this ref:
```python
@_make_elementwise_unary_reference(ELEMENTWISE_TYPE_PROMOTION_KIND.ALWAYS_BOOL)
def isposinf(a: TensorLikeType) -> TensorLikeType:
    if a.dtype.is_complex:
        msg = f"Complex dtype is not supported for isposinf, got dtype {a.dtype}"
        raise ValueError(msg)
    elif a.dtype.is_floating_point:
        return eq(a, float("inf"))
    return zeros_like(a, dtype=torch.bool)
```
where for this case, it is just a `return constant_tensor`.

But as you have pointed out in https://github.com/pytorch/pytorch/pull/78655/files#r899877833, this is not sufficient to handle constant tensors for all cases.

File: torch/_prims/executor.py
Comment (zasdfgbnm) [reply torch/_prims/executor.py:76]: Yes, there might be constant tensors here.

File: torch/_prims/executor.py
Comment (zasdfgbnm) [reply torch/_prims/executor.py]: IIRC, when I wrote this code, there was no such `_to_nvfuser_constant`, so seems that we did the same work in different ways.

Comment (zasdfgbnm) [reply torch/_prims/executor.py]: I actually need to keep the `_to_nvfuser_constant` instead of using the `to_nv`, because according to https://github.com/pytorch/pytorch/pull/78655#discussion_r899858111, `to_nv` will convert numbers as scalar, and we will need additional inputs to the fusion.

File: torch/_prims/executor.py
Comment (zasdfgbnm) [reply torch/_prims/executor.py]: done

File: torch/_prims/executor.py
Comment (zasdfgbnm) [reply torch/_prims/executor.py]: done

File: torch/_prims/executor.py
Comment (zasdfgbnm) [reply torch/_prims/executor.py]: I reverted this change, see https://github.com/pytorch/pytorch/pull/78655#discussion_r900516154

File: torch/_prims/executor.py
Comment (zasdfgbnm) [reply torch/_prims/executor.py]: Not sure if it is a good idea to do so, but I added a new commit https://github.com/pytorch/pytorch/pull/78655/commits/df55ea7a608363eb2290fcafda854d864bb72f7b to handle constant tensor by making it a fusion input.

**Review by IvanYashchuk — APPROVED** [review]
`torch/_prims/executor.py` changes look good to me. Thank you!

**Comment by zasdfgbnm** [comment]
I am seeing
```
======================================================================
ERROR: test_python_ref_meta__refs_isinf_cpu_int8 (__main__.TestCommonCPU)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/gaoxiang/.local/lib/python3.10/site-packages/torch/testing/_internal/common_device_type.py", line 377, in instantiated_test
    result = test(self, **param_kwargs)
  File "/home/gaoxiang/.local/lib/python3.10/site-packages/torch/testing/_internal/common_device_type.py", line 786, in test_wrapper
    return test(*args, **kwargs)
  File "/home/gaoxiang/.local/lib/python3.10/site-packages/torch/testing/_internal/common_device_type.py", line 979, in only_fn
    return fn(self, *args, **kwargs)
  File "/home/gaoxiang/pytorch-prims/test/test_ops.py", line 379, in test_python_ref_meta
    meta_result = op(meta_sample.input, *meta_sample.args, **meta_sample.kwargs)
  File "/home/gaoxiang/.local/lib/python3.10/site-packages/torch/testing/_internal/common_methods_invocations.py", line 955, in __call__
    return self.op(*args, **kwargs)
  File "/home/gaoxiang/.local/lib/python3.10/site-packages/torch/_prims/wrappers.py", line 173, in _fn
    result = fn(*args, **kwargs)
  File "/home/gaoxiang/.local/lib/python3.10/site-packages/torch/_prims/wrappers.py", line 263, in _fn
    return fn(*args, **kwargs)
  File "/home/gaoxiang/.local/lib/python3.10/site-packages/torch/_prims/wrappers.py", line 114, in _fn
    result = fn(**bound.arguments)
  File "/home/gaoxiang/.local/lib/python3.10/site-packages/torch/_refs/__init__.py", line 343, in _ref
    return prim(a)
  File "/home/gaoxiang/.local/lib/python3.10/site-packages/torch/_refs/__init__.py", line 491, in isinf
    return logical_or(isposinf(a), isposinf(a))
  File "/home/gaoxiang/.local/lib/python3.10/site-packages/torch/_prims/wrappers.py", line 173, in _fn
    result = fn(*args, **kwargs)
  File "/home/gaoxiang/.local/lib/python3.10/site-packages/torch/_prims/wrappers.py", line 114, in _fn
    result = fn(**bound.arguments)
  File "/home/gaoxiang/.local/lib/python3.10/site-packages/torch/_refs/__init__.py", line 754, in _ref
    return prim(a, b)
  File "/home/gaoxiang/.local/lib/python3.10/site-packages/torch/_refs/__init__.py", line 1061, in _logical_or
    return a | b
  File "/home/gaoxiang/.local/lib/python3.10/site-packages/torch/utils/_python_dispatch.py", line 74, in wrapped
    return f(self, *args, **kwargs)
  File "/home/gaoxiang/.local/lib/python3.10/site-packages/torch/_subclasses/fake_tensor.py", line 440, in __torch_dispatch__
    raise Exception(
Exception: Invoking operators with non-Fake Tensor inputs in FakeTensorMode is not yet supported. Please convert all Tensors to FakeTensors first. Found in aten.bitwise_or.Tensor

----------------------------------------------------------------------
```
with this PR, and after changing
```
def _logical_or(a: TensorLikeType, b: TensorLikeType):
    if not utils.is_boolean_dtype(a.dtype):
        a = a != 0
    if not utils.is_boolean_dtype(b.dtype):
        b = b != 0
    return a | b
    return bitwise_or(a, b)
```
to
```
def _logical_or(a: TensorLikeType, b: TensorLikeType):
    if not utils.is_boolean_dtype(a.dtype):
        a = a != 0
    if not utils.is_boolean_dtype(b.dtype):
        b = b != 0
    return bitwise_or(a, b)
```
the problem is resolved. @Chillee Do you have an idea of what is happening here?

**Comment by Chillee** [comment]
Do you see an error if you use `torch.bitwise_or`?

**Comment by zasdfgbnm** [comment]
> Do you see an error if you use `torch.bitwise_or`?

Yes, I see the same error.

**Comment by Chillee** [comment]
Seems like some kind of issue with FakeTensor (perhaps `logical_or` is getting a non-Faketensor and a real tensor?). 

For now, switching it back to `bitwise_or` is fine to unblock you.

cc: @eellison

**Comment by mruberry** [comment]
Please file and comment an issue if you have to workaround using the torch op, @zasdfgbnm

**Pull request closed without merging**
